File: frontend/lib/backend/agents/credit-risk-agent.ts
import type { AgentReport, Applicant, Contributor } from "../schemas";

export function runCreditRiskAgent(applicant: Applicant): AgentReport {
  const signals = applicant.document_signals ?? {};
  const numericHints = signals.numeric_hints ?? {};

  const effectiveUtilization = Math.max(
    applicant.credit_utilization,
    numericHints.credit_utilization ?? applicant.credit_utilization
  );
  const effectiveDelinquencies = Math.max(
    applicant.delinquencies_12m,
    numericHints.delinquencies_12m ?? applicant.delinquencies_12m
  );
  const effectiveEmploymentMonths = Math.min(
    applicant.employment_months,
    numericHints.employment_months ?? applicant.employment_months
  );

  let risk = 0.1;
  risk += effectiveUtilization * 0.35;
  risk += Math.min(effectiveDelinquencies, 4) * 0.12;
  risk +=
    effectiveEmploymentMonths < 6
      ? 0.15
      : effectiveEmploymentMonths < 12
        ? 0.07
        : -0.05;
  risk +=
    Math.min(
      applicant.existing_debt / Math.max(applicant.monthly_income * 12, 1),
      1.5
    ) * 0.15;

  const score = Math.max(0, Math.min(1, 1 - risk));

  const contributors: Contributor[] = [
    {
      feature: "credit_utilization",
      impact: Math.round(effectiveUtilization * 0.35 * 100) / 100,
      direction: "increases risk",
      explanation: `Credit utilization is ${(effectiveUtilization * 100).toFixed(0)}%.`,
    },
    {
      feature: "delinquencies_12m",
      impact: Math.round(Math.min(effectiveDelinquencies, 4) * 0.12 * 100) / 100,
      direction: "increases risk",
      explanation: `Recent delinquencies: ${effectiveDelinquencies}.`,
    },
    {
      feature: "employment_stability",
      impact: effectiveEmploymentMonths >= 12 ? -0.05 : 0.15,
      direction:
        effectiveEmploymentMonths >= 12 ? "decreases risk" : "increases risk",
      explanation: `Employment history: ${effectiveEmploymentMonths} months.`,
    },
  ];

  const reasons: string[] = [];
  if (effectiveUtilization > 0.75) reasons.push("HIGH_UTILIZATION");
  if (effectiveDelinquencies > 0) reasons.push("RECENT_DELINQUENCY");
  if (effectiveEmploymentMonths < 6) reasons.push("SHORT_EMPLOYMENT_HISTORY");
  if (Object.keys(numericHints).length) reasons.push("DOC_HINTS_APPLIED");
  if (!reasons.length) reasons.push("STABLE_CREDIT_PROFILE");

  const recommendation =
    score >= 0.74 ? "PASS" : score >= 0.45 ? "REFER" : "REJECT";

  return {
    agent_name: "Credit Risk Agent",
    score: Math.round(score * 100) / 100,
    model_version: "credit-risk-monotonic-xgb-v1",
    top_contributors: contributors,
    evidence_refs: ["credit_bureau_snapshot"],
    reason_codes: reasons,
    confidence_status: "demo_calibrated",
    recommendation,
    summary:
      "Probability-of-default style model with TreeSHAP-like reason contributions.",
  };
}

File: backend/app/agents/credit_risk_agent.py
from app.models.schemas import Applicant, AgentReport, Contributor

def run(applicant: Applicant) -> AgentReport:
    signals = applicant.document_signals or {}
    numeric_hints = signals.get("numeric_hints", {})

    effective_utilization = max(applicant.credit_utilization, float(numeric_hints.get("credit_utilization", applicant.credit_utilization)))
    effective_delinquencies = max(applicant.delinquencies_12m, int(numeric_hints.get("delinquencies_12m", applicant.delinquencies_12m)))
    effective_employment_months = min(applicant.employment_months, int(numeric_hints.get("employment_months", applicant.employment_months)))

    risk = 0.10
    risk += effective_utilization * 0.35
    risk += min(effective_delinquencies, 4) * 0.12
    risk += 0.15 if effective_employment_months < 6 else 0.07 if effective_employment_months < 12 else -0.05
    risk += min(applicant.existing_debt / max(applicant.monthly_income*12, 1), 1.5) * 0.15
    score = max(0.0, min(1.0, 1-risk))

    contributors=[
        Contributor(feature="credit_utilization", impact=round(effective_utilization*0.35,2), direction="increases risk", explanation=f"Credit utilization is {effective_utilization:.0%}."),
        Contributor(feature="delinquencies_12m", impact=round(min(effective_delinquencies,4)*0.12,2), direction="increases risk", explanation=f"Recent delinquencies: {effective_delinquencies}."),
        Contributor(feature="employment_stability", impact=-0.05 if effective_employment_months>=12 else 0.15, direction="decreases risk" if effective_employment_months>=12 else "increases risk", explanation=f"Employment history: {effective_employment_months} months."),
    ]
    reasons=[]
    if effective_utilization > .75: reasons.append("HIGH_UTILIZATION")
    if effective_delinquencies > 0: reasons.append("RECENT_DELINQUENCY")
    if effective_employment_months < 6: reasons.append("SHORT_EMPLOYMENT_HISTORY")
    if numeric_hints:
        reasons.append("DOC_HINTS_APPLIED")
    if not reasons: reasons=["STABLE_CREDIT_PROFILE"]
    recommendation = "PASS" if score >= .74 else "REFER" if score >= .45 else "REJECT"
    return AgentReport(agent_name="Credit Risk Agent", score=round(score,2), model_version="credit-risk-monotonic-xgb-v1", top_contributors=contributors, evidence_refs=["credit_bureau_snapshot"], reason_codes=reasons, confidence_status="demo_calibrated", recommendation=recommendation, summary="Probability-of-default style model with TreeSHAP-like reason contributions.")
